### PR TITLE
Restart sumo process only when map file changes

### PR DESCRIPTION
**Pull request by liamchzh**
SMARTS forces the reinitialization of TRACI and the SUMO process to prevent the simulation from freezing. Though this change causes overhead costs, especially when switching between large maps, it solved the problem. 

A better way to decide whether to restart the SUMO process or not is based on the map. Ideally, we restart SUMO only if the map changes, otherwise, just reset everything.

### CODE REVIEW

File: smarts/core/sumo_traffic_simulation.py [hunk 1]
```diff
@@ -266,7 +271,6 @@ def teardown(self):
             return
 
         assert self._is_setup
-        self._close_traci_and_pipes()
```
Comment (Gamenot) [inline smarts/core/sumo_traffic_simulation.py:269]: If you are removing the traci close guarantee here I would suggest adding a destroy method to make sure that the pipes get closed.

**Review by Gamenot — CHANGES_REQUESTED** [review]

File: smarts/core/sumo_traffic_simulation.py (hunk 1 above)
Comment (liamchzh) [reply smarts/core/sumo_traffic_simulation.py:269]: Actually `initialize_traci_conn` will call `_close_traci_and_pipes` so I think the pipes are not leaking.

**Comment by JVillella** [comment]
Just write as

```python
# restart sumo process only when map file changes
if self._scenario and self._scenario.net_filepath == scenario.net_filepath:
    self._traci_conn.load(self._base_sumo_load_params())
else:
    self._initialize_traci_conn()
```
?

**Review by JVillella — APPROVED** [review]
Seems fine 👍 (btw, can we use PR description to provide context?)

**Comment by liamchzh** [comment]
> Seems fine  (btw, can we use PR description to provide context?)

For sure.

**Comment by liamchzh** [comment]
> Just write as
> 
> ```python
> # restart sumo process only when map file changes
> if self._scenario and self._scenario.net_filepath == scenario.net_filepath:
>     self._traci_conn.load(self._base_sumo_load_params())
> else:
>     self._initialize_traci_conn()
> ```

Now I remember why I need a `restart_sumo`, because the `self._scenario` needs to be assigned first, so either an extra variable, 

or 
```
        if self._scenario and self._scenario.net_filepath == scenario.net_filepath:
            self._scenario = scenario
            self._traci_conn.load(self._base_sumo_load_params())
        else:
            self._scenario = scenario
            self._initialize_traci_conn()
```

**Review by Gamenot — APPROVED** [review]

**Pull request merged**